### PR TITLE
Integrate Register form

### PR DESCRIPTION
**Pull request by subinasr**
- Make four different registration forms
- Make individual user form work
- Make institution user form work

### CODE REVIEW

File: app/Base/components/Navbar/index.tsx [hunk 1]
```diff
@@ -116,12 +116,12 @@ function Navbar(props: Props) {
                 </div>
                 {!(authenticated && user) && (
                     <div className={styles.actions}>
-                        <Button
-                            name={undefined}
-                            onClick={undefined}
+                        <ButtonLikeLink
```
Comment (tnagorra) [inline app/Base/components/Navbar/index.tsx:119]: Use SmartLink if possible. No need to check for authenticated and user.

File: app/views/Register/RegisterForm/IndividualUserForm/index.tsx [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+            profile: $profile,
+            phoneNumber: $phoneNumber,
+        }) {
+            errors
+            ok
+        }
+    }
+`;
+
+interface IndividualRegistrationFields {
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/IndividualUserForm/index.tsx]: Why not use generated types?

File: app/views/Register/RegisterForm/IndividualUserForm/index.tsx [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+            onCompleted: (response) => {
+                console.info('Registeration completed', response);
+                const { register: registerRes } = response;
+                if (!register) {
+                    return;
+                }
+                const {
+                    ok,
+                    errors,
+                } = registerRes ?? {};
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/IndividualUserForm/index.tsx]: Do not set `{}` as default value to destructure variables

File: app/views/Register/RegisterForm/IndividualUserForm/index.tsx [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+                firstName: finalValue?.firstName,
+                lastName: finalValue?.lastName,
+                phoneNumber: finalValue?.phoneNumber,
+                profile: {},
+                userType: 'INDIVIDUAL_USER',
+            },
+        });
+    }, []);
+
+    const confirmationError = React.useMemo(() => {
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/IndividualUserForm/index.tsx]: Look like this needs more work

File: app/views/Register/RegisterForm/IndividualUserForm/index.tsx [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+            className={styles.registerForm}
+            onSubmit={createSubmitHandler(validate, setError, handleSubmit)}
+        >
+            <TextInput
+                name="email"
+                label="Email"
+                value={value?.email}
+                error={error?.email}
+                onChange={setFieldValue}
+                placeholder="johndoe@email.com"
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/IndividualUserForm/index.tsx]: Let's remove placeholders

File: app/views/Register/RegisterForm/InstitutionForm/index.tsx [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+    ) {
+        register(data: {
+            email: $email,
+            password: $password,
+            firstName: $firstName,
+            lastName: $lastName,
+            userType: $userType,
+            phoneNumber: $phoneNumber,
+            profile: {
+                institution: {
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/InstitutionForm/index.tsx]: Always fetch `id` for all nodes requested from gql query/mutation

File: app/views/Register/RegisterForm/InstitutionForm/index.tsx [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+            districts: $districtIds,
+        ) {
+            results {
+                id
+                name
+            }
+            totalCount
+        }
+    }
+`;
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/InstitutionForm/index.tsx]: We should fetch these lazily

File: app/views/Register/RegisterForm/InstitutionForm/index.tsx [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+
+interface InstitutionalRegistrationFields {
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/InstitutionForm/index.tsx]: Why not use generated types for form fields?

File: app/views/Register/RegisterForm/InstitutionForm/index.tsx [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+    const {
+        data: provinceList,
+        loading: provincesPending,
+    } = useQuery<ProvinceListQuery, ProvinceListQueryVariables>(
+        PROVINCE_LIST,
+    );
+
+    const districtVariables = useMemo(() => ({
+        provinceIds: [value?.province],
+    }), []);
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/InstitutionForm/index.tsx]: The dependency array is missing value?.province

File: app/views/Register/RegisterForm/InstitutionForm/index.tsx [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+        DISTRICT_LIST,
+        {
+            variables: districtVariables,
+        },
+    );
+
+    const municipalityVariables = useMemo(() => ({
+        provinceIds: [value?.province],
+        districtIds: [value?.district],
+    }), []);
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/InstitutionForm/index.tsx]: The dependency array is missing values

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/views/Register/RegisterForm/index.tsx [hunk 11]
```diff
@@ -1,135 +1,274 @@
...
+            case 'INSTITUTIONAL_USER':
+                delete finalValues.school;
+                delete finalValues.publisher;
+                break;
+            default:
+                delete finalValues.school;
+                delete finalValues.institution;
+                delete finalValues.publisher;
+                break;
+        }
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/index.tsx]: Let's define this dynamically on schema

File: app/views/Register/RegisterForm/index.tsx [hunk 12]
```diff
@@ -1,135 +1,274 @@
...
+                }
 
-    const handleSubmit = useCallback(() => {
-        console.warn('to be handled');
-        // TODO: handle Submit
-    }, []);
+                const {
+                    ok,
+                    errors,
+                } = registerResponse ?? {};
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/index.tsx]: Let's not do this

File: app/views/Register/RegisterForm/index.tsx [hunk 13]
```diff
@@ -1,135 +1,274 @@
...
-            lengthGreaterThanCondition(4),
-            lengthSmallerThanCondition(129),
-        ],
-    }),
+interface UserType {
+    id: UserUserType;
+    title: string;
+}
+
+const userType: UserType[] = [
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/index.tsx:48]: We should be able to fetch these from server

File: app/views/Register/RegisterForm/common.ts [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    email: string;
+    firstName?: string;
+    lastName?: string;
+    password: string;
+    userType: UserUserType;
+    phoneNumber: string;
+    institution?: InstitutionFields;
+    publisher?: PublisherFields;
+    school?: SchoolFields;
+}
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/common.ts]: Let's use type from server

File: app/views/Register/RegisterForm/PublisherForm/index.tsx [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    error: Error<PublisherType>;
+    onChange: (value: SetValueArg<PublisherInputValue> | undefined, name: K) => void;
+}
+
+function PublisherForm<K extends string>(props: Props<K>) {
+    const {
+        name,
+        value,
+        error: formError,
+        onChange,
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/PublisherForm/index.tsx:35]: Let's add disabled prop and disable these inputs when form is being submitted

File: app/views/Register/RegisterForm/LocationInput/index.tsx [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+        districtIds: district ? [district] : undefined,
+    }), [province, district]);
+
+    const {
+        data: municipalityList,
+        loading: municipalitiesPending,
+    } = useQuery<MunicipalityListQuery, MunicipalityListQueryVariables>(
+        MUNICIPALITY_LIST,
+        { variables: municipalityVariables },
+    );
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/LocationInput/index.tsx]: Let's not pull all of these at once

File: app/views/Register/RegisterForm/LocationInput/index.tsx [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+
+    const [province, setProvince] = useInputState<Province['id'] | undefined>(undefined);
+    const [district, setDistrict] = useInputState<District['id'] | undefined>(undefined);
+    const [municipality, setMunicipality] = useInputState<Municipality['id'] | undefined>(undefined);
+
+    React.useEffect(() => {
+        if (onChange) {
+            onChange(municipality, name);
+        }
+    }, [name, onChange, municipality]);
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/LocationInput/index.tsx]: Let's not use useEffect

File: app/views/Register/RegisterForm/LocationInput/index.tsx [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+
+    React.useEffect(() => {
+        setMunicipality(value);
+    }, [value, setMunicipality]);
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/LocationInput/index.tsx]: Let's not use useEffect

File: app/views/Register/RegisterForm/IndividualUserForm/index.tsx [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+                        {
+                            variant: 'error',
+                        },
+                    );
+                }
+            },
+        },
+    );
+
+    const handleSubmit = useCallback((finalValue) => {
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/IndividualUserForm/index.tsx]: Let's define type for finalValue

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/views/Register/RegisterForm/common.ts [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
```
Comment (tnagorra) [inline app/views/Register/RegisterForm/common.ts:42]: Let's use interfaces from server

File: app/views/Register/RegisterForm/index.tsx (hunk 13 above)
Comment (frozenhelium) [reply app/views/Register/RegisterForm/index.tsx:48]: Lets leave it as is for now

File: app/views/Register/RegisterForm/common.ts (hunk 20 above)
Comment (frozenhelium) [reply app/views/Register/RegisterForm/common.ts:42]: It's slightly different from the current usecase

**Pull request merged**